ntcore/src/main/native/include/networktables/ValueListener.h

```diff
@@ -0,0 +1,204 @@
+   * @param listener Listener function
+   */
+  ValueListener(NT_Handle subentry, unsigned int mask,
+                std::function<void(const ValueNotification&)> listener);
+
+  ValueListener(const ValueListener&) = delete;
+  ValueListener& operator=(const ValueListener&) = delete;
+  ValueListener(ValueListener&& rhs);
+  ValueListener& operator=(ValueListener&& rhs);
+  ~ValueListener();
+
+  explicit operator bool() const { return m_handle != 0; }
+
+  /**
+   * Gets the native handle.
+   *
+   * @return Handle
+   */
+  NT_ValueListener GetHandle() const { return m_handle; }
+
+ private:
+  NT_ValueListener m_handle{0};
+};
+
+/**
+ * Value change listener. This queues value change events matching the specified
+ * mask. Code using the listener must periodically call readQueue() to read the
+ * events.
+ */
+class ValueListenerPoller final {
+ public:
+  ValueListenerPoller();
+
+  /**
+   * Construct a value listener poller.
+   *
+   * @param inst Instance
+   */
+  explicit ValueListenerPoller(NetworkTableInstance inst);
+
+  ValueListenerPoller(const ValueListenerPoller&) = delete;
+  ValueListenerPoller& operator=(const ValueListenerPoller&) = delete;
+  ValueListenerPoller(ValueListenerPoller&& rhs);
+  ValueListenerPoller& operator=(ValueListenerPoller&& rhs);
+  ~ValueListenerPoller();
+
+  explicit operator bool() const { return m_handle != 0; }
+
+  /**
+   * Gets the native handle.
+   *
+   * @return Handle
+   */
+  NT_ValueListenerPoller GetHandle() const { return m_handle; }
+
+  /**
+   * Start listening to value changes on a subscriber.
+   *
+   * @param subscriber Subscriber
+   * @param mask Bitmask of ValueListenerFlags values
+   * @return Listener handle
+   */
+  NT_ValueListener Add(const Subscriber& subscriber, unsigned int mask);
+
+  /**
+   * Start listening to value changes on a subscriber.
+   *
+   * @param subscriber Subscriber
+   * @param mask Bitmask of ValueListenerFlags values
+   * @return Listener handle
+   */
+  NT_ValueListener Add(const MultiSubscriber& subscriber, unsigned int mask);
+
+  /**
+   * Start listening to value changes on an entry.
+   *
+   * @param entry Entry
+   * @param mask Bitmask of ValueListenerFlags values
+   * @return Listener handle
+   */
+  NT_ValueListener Add(const NetworkTableEntry& entry, unsigned int mask);
+
+  /**
+   * Start listening to value changes on a subscriber/entry handle.
+   *
+   * @param subentry Subscriber/entry handle
+   * @param mask Bitmask of ValueListenerFlags values
+   * @return Listener handle
+   */
+  NT_ValueListener Add(NT_Handle subentry, unsigned int mask);
+
+  /**
+   * Remove a listener.
+   *
+   * @param listener Listener handle
+   */
+  void Remove(NT_ValueListener listener);
+
+  /**
+   * Reads value listener queue (all value changes since last call).
+   *
+   * @param poller    poller handle
+   * @return Array of value notifications.
+   */
+  std::vector<ValueNotification> ReadQueue();
```

Comment:
Documentation should clarify whether you see intermediate changes (x=1, x=2 .. x=n) or if the changes are rolled up (x=n only).

Reply:
We say that we queue value change events matching the specified mask.  I'm not sure how that could be interpreted as losing value changes, or how better to say that?